packages/contracts-bedrock/src/cannon/libraries/MIPSMemory.sol

```diff
@@ -0,0 +1,116 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+library MIPSMemory {
+    /// @notice Reads a 32-bit value from memory.
+    /// @param _memRoot The current memory root
+    /// @param _addr The address to read from.
+    /// @param _proofOffset The offset of the memory proof in calldata.
+    /// @return out_ The hashed MIPS state.
+    function readMem(bytes32 _memRoot, uint32 _addr, uint256 _proofOffset) internal pure returns (uint32 out_) {
+        unchecked {
+            assembly {
+                // Validate the address alignement.
+                if and(_addr, 3) { revert(0, 0) }
+
+                // Load the leaf value.
+                let leaf := calldataload(_proofOffset)
```

Comment:
It's worth asserting here again (and in `writeMem`) that there's sufficient calldata for the proof. (see https://github.com/ethereum-optimism/optimism/pull/10978/files#diff-2543eb93ac56f9072b671119906a62fec2a87591132b4315ef4745b84c8c43cbR112 for example).

Otherwise, misuse of this function will be silent as calling `calldataoad` for offsets beyond actual calldata returns zero.

Reply:
Good catch - I moved the calldata length validation from `memoryProofOffset` into `readMem` and `writeMem`.